netdata/netdata
# Port the OpenLDAP collector from Python to Go

##### Summary

Port the Python implementation of OpenLDAP to Go

@ilyam8 the new module added some more stuff, check and let me know if I need to do anything

##### Test Plan

1. Install OpenLDAP
	First, install the necessary OpenLDAP packages on your system:
	
	```
	sudo apt update  
	sudo apt install slapd ldap-utils
	```	

	During the installation process, you may be prompted to set the admin password for LDAP. Make sure to remember this password for later use.

2. Configure OpenLDAP
	After installation, reconfigure OpenLDAP to set up your domain:

	```
	sudo dpkg-reconfigure slapd
	```

	Follow the prompts:
	- Omit OpenLDAP server configuration? No
	- DNS domain name: (Set your domain, e.g., example.com)
	- Organization name: (Set your organization name, e.g., Example Corp)
	- Administrator password: (Enter the admin password)
	- Database backend to use: MDB
	- Remove the database when slapd is purged? No
	- Move old database? Yes

3. Add Groups and Users
	Create `group.ldif` with the following content:
	
	```
	dn: cn=developers,ou=groups,dc=example,dc=com  
	objectClass: top  
	objectClass: posixGroup  
	cn: developers  
	gidNumber: 10000
	```

	Apply the group configuration:
	
	```	
	ldapadd -x -D "cn=admin,dc=example,dc=com" -W -f group.ldif
	```	

	Create `user.ldif` with the following content:

	```	
	dn: uid=john,ou=users,dc=example,dc=com  
	objectClass: inetOrgPerson  
	objectClass: posixAccount  
	objectClass: shadowAccount  
	cn: John Doe  
	sn: Doe  
	uid: john  
	uidNumber: 10001  
	gidNumber: 10000  
	homeDirectory: /home/john  
	loginShell: /bin/bash  
	userPassword: {SSHA}password_hash_here
	```

	Apply the user configuration:
	
	```	
	ldapadd -x -D "cn=admin,dc=example,dc=com" -W -f user.ldif
	```

4. Enable Monitoring Backend

	First, create a file `monitor.ldif` with the following content:
	
	```	
	dn: cn=module{0},cn=config  
	changetype: modify  
	add: olcModuleLoad  
	olcModuleLoad: back_monitor.la
	```

	Apply the configuration:

	```	
	ldapmodify -Y EXTERNAL -H ldapi:/// -f monitor.ldif
	```

	Now, add the monitoring database:

	```	
	dn: olcDatabase=Monitor,cn=config  
	objectClass: olcDatabaseConfig  
	olcDatabase: Monitor
	```

	Apply it:
	
	```	
	ldapadd -Y EXTERNAL -H ldapi:/// -f monitor.ldif
	```

5. Verify Setup

	```
	ldapsearch -x -LLL -b dc=example,dc=com
	```
   	
	You should see your users, groups, and monitor backend listed.

6. Command
	
	An equivalent of the command that the package uses is:

	```
	ldapsearch -x -D "cn=admin,dc=example,dc=com" -W -b "cn=Total,cn=Connections,cn=Monitor" -s base monitorCounter

	```
